### PR TITLE
Added method to recover all registered WA on the connected UA

### PR DESCRIPTION
**Pull request by frsela**

### CODE REVIEW

**Comment by frsela** [comment]
Related issue #115
@acperez could you test this too?

**Comment by acperez** [comment]
Tested but not working, I get error 450 from server:

I/Gecko   (  109): -_\- PushNotificationService: registerUA::start
I/Gecko   (  109): -_\- PushNotificationService: registerUA::createWS ws://192.168.43.252:8080
I/Gecko   (  109): -_\- PushNotificationService: registerUA::onStart
I/Gecko   (  109): -_\- PushNotificationService: registerUA: {"messageType":"registerUA","data":{"uatoken":"ff54b61c17de76062d7b96f4ecba7e1375b8afcc406d8310972c48acf753284121e0c581f38df92b6be840dc60f6c15702e8db777f8980676d5c50a293ef66c0d562519b98a0a28a2af7d08395d035545907f09023410165b095a9377b4fe7403f7d0f2a35fe976a1dc194a2ec98883f1a1e4e6e4b97e016702e6d7170e0baa2c415b9004cdb50f6f162080c924d9b5c","interface":{},"mobilenetwork":{"mcc":0,"mnc":0}}}
I/Gecko   (  109): -_\- PushNotificationService: Send message and set timer
I/Gecko   (  109): -_\- PushNotificationService: onAcknowledge 387
I/Gecko   (  109): -_\- PushNotificationService: Cancel send timer
I/Gecko   (  109): -_\- PushNotificationService: message: {"messageType":"registerUA","status":"REGISTERED","statuscode":200}
I/Gecko   (  109): -_\- PushNotificationService: registerUA::onMessageAvailable
I/Gecko   (  109): -_\- PushNotificationService: getRegisteredWa::start
I/Gecko   (  109): -_\- PushNotificationService: ------------------------------------
I/Gecko   (  109): -_\- PushNotificationService: getRegisteredWAs: {"messageType":"getRegisteredWA"}
I/Gecko   (  109): -_\- PushNotificationService: Send message and set timer
I/Gecko   (  109): -_\- PushNotificationService: onAcknowledge 6
I/Gecko   (  109): -_\- PushNotificationService: Cancel send timer
I/Gecko   (  109): -_\- PushNotificationService: message: {"statuscode":450,"status":"ERROR","reason":"Data received is not a valid JSON package"}

I validated JSON online and it seems to be valid.
{"messageType":"getRegisteredWA"}

**Comment by frsela** [comment]
Do you've the server traces? I'm testing here and it's working well :s

File: src/common/datastore.js [hunk 1]
```diff
@@ -196,6 +196,29 @@ var DataStore = function() {
   },
 
   /**
+   * Recover a list of WA associated to a UA
+   */
+  this.getApplicationsOnUA = function (uaToken, callbackFunc, callbackParam) {
```
Comment (willyaranda) [inline src/common/datastore.js]: nit: `getApplicationsForUA`

File: src/common/datastore.js [hunk 2]
```diff
@@ -196,6 +196,29 @@ var DataStore = function() {
...
+    // Get from MongoDB
+    log.debug("datastore::getApplicationsOnUA --> Going to find applications in UA: " + uaToken);
+    this.db.collection("apps", function(err, collection) {
+      if (!err) {
+        collection.find( { node: uaToken }, { _id:"" } ).toArray(function(err,d) {
```
Comment (willyaranda) [inline src/common/datastore.js]: question: I can understand why we are finding for `node: uaToken` but not the `{ _id:""}` part. I can see the API here: https://github.com/mongodb/node-mongodb-native/blob/master/docs/queries.md , but I'm a bit confused.

File: src/ns_ua/datamanager.js [hunk 3]
```diff
@@ -139,6 +139,15 @@ datamanager.prototype = {
   },
 
   /**
+   * Recover a list of WA associated to a UA
+   */
+  getApplicationsOnUA: function (uaToken, callback) {
```
Comment (willyaranda) [inline src/ns_ua/datamanager.js]: Same: `getApplicationsForUA`

File: src/ns_ua/ws_server.js [hunk 4]
```diff
@@ -434,6 +434,43 @@ server.prototype = {
...
 
+          case "getRegisteredWA":
+            log.debug("WS::onWSMessage::getRegisteredWA --> Recovering list of registered WA");
+
+            if(!dataManager.getUAToken(connection)) {
+              log.debug("No UAToken found for this connection !");
+              connection.res({
+                errorcode: errorcodes.UATOKEN_NOTFOUND,
+                extradata: {
+                  'WATokens': {},
```
Comment (willyaranda) [inline src/ns_ua/ws_server.js]: object or array? I'd prefer an array (easier to parse later on response)

File: src/ns_ua/ws_server.js [hunk 5]
```diff
@@ -434,6 +434,43 @@ server.prototype = {
...
+              break;
+            }
+            dataManager.getApplicationsOnUA(dataManager.getUAToken(connection),
+              function (d) {
+                log.debug("",d);
+                if(!d) {
+                  connection.res({
+                    errorcode: errorcodes.NO_ERROR,
+                    extradata: {
+                      'WATokens': {},
```
Comment (willyaranda) [inline src/ns_ua/ws_server.js]: again, I'd prefer array

File: src/ns_ua/ws_server.js (hunk 4 above)
Comment (willyaranda) [inline src/ns_ua/ws_server.js]: In fact: https://github.com/telefonicaid/notification_server/pull/119/files#L0R206 returns an array, not an object ;)

**Comment by willyaranda** [comment]
Albert, can you send the traces as Fernando said? I cannot see the problem with this data.

File: src/common/datastore.js (hunk 1 above)
Comment (frsela) [inline src/common/datastore.js]: agree

File: src/common/datastore.js (hunk 2 above)
Comment (frsela) [inline src/common/datastore.js]: well, the _id is basically the apptoken ...
Probably will be better to respond the URL and not only the apptoken.

Comment (frsela) [inline src/common/datastore.js]: Oh, about your doubt.

I recover all apps which have the uaToken in theirs node array ... and only recover the _id field.

File: src/ns_ua/ws_server.js (hunk 4 above)
Comment (frsela) [inline src/ns_ua/ws_server.js]: Array, it's my fault .... :p

File: src/common/datastore.js (hunk 2 above)
Comment (willyaranda) [inline src/common/datastore.js]: so, this is the filter you are passing to the function, right? that means that you want to get only the `_id`, right? I think, we can get, as you said, both things: the `_id` and the `URL`

Comment (willyaranda) [inline src/common/datastore.js]: BTW, the doc sais this:

To indicate which fields must or must no be returned fields value can be used. For example the following fields value

```
{
    "name": true,
    "title": true
}
```

so you need to put: `{ _id: true, URL: true }`

Comment (frsela) [inline src/common/datastore.js]: The URL is not stored in mongo, should be generated ... or stored :s

Comment (willyaranda) [inline src/common/datastore.js]: So, let's send right now just the WAtoken, we can think if storing and sending the URL has some benefits.

**Comment by acperez** [comment]
@willyaranda Problem solved, was my fault when sending the message

**Pull request merged**